assertion failed: unresolved symbols while pickling quote with type bounds

## Compiler version

- 3.1.2-RC2
- 3.1.1
- 3.1.3-RC1-bin-20220317-17e46ad-NIGHTLY


## Minimized code


```Scala
import scala.quoted.*

object Main {
  def foo(a: Expr[Any])(using Quotes) = {
    a match {
      case '{ ($x: Set[t]).toSet } =>
      case _ =>
    }
  }
}
```

### workaround(?)

```scala
import scala.quoted.*

object Main {
  def foo(a: Expr[Any])(using Quotes) = {
    a match {
      case '{
              type t1
              type t2 >: `t1`
              ($x: Set[`t1`]).toSet[`t2`]
            } =>
      case _ =>
    }
  }
}
```

## Output (click arrow to expand)
<details>

```
[error] java.lang.AssertionError: assertion failed: unresolved symbols: value t$given1 (line 5) #14166 when pickling /Users/kenji/set-expr-match/Main.scala
[error] scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickle(TreePickler.scala:777)
[error] dotty.tools.dotc.transform.Pickler.run$$anonfun$1$$anonfun$1(Pickler.scala:69)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.transform.Pickler.run$$anonfun$1(Pickler.scala:106)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.transform.Pickler.run(Pickler.scala:106)
[error] dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:308)
[error] scala.collection.immutable.List.map(List.scala:246)
[error] dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:309)
[error] dotty.tools.dotc.transform.Pickler.runOn(Pickler.scala:111)
[error] dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:261)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error] dotty.tools.dotc.Run.runPhases$1(Run.scala:272)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:280)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:68)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:289)
[error] dotty.tools.dotc.Run.compileSources(Run.scala:222)
[error] dotty.tools.dotc.Run.compile(Run.scala:206)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
[error] dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:88)
[error] dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:192)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:247)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:182)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:163)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:163)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:210)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:177)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:175)
[error] sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:461)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:416)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:503)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:403)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:169)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:482)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2366)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2316)
[error] sbt.internal.server.BspCompileTask$.$anonfun$compute$1(BspCompileTask.scala:30)
[error] sbt.internal.io.Retry$.apply(Retry.scala:46)
[error] sbt.internal.io.Retry$.apply(Retry.scala:28)
[error] sbt.internal.io.Retry$.apply(Retry.scala:23)
[error] sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:30)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2314)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] sbt.Execute.work(Execute.scala:291)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error] java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error] java.lang.Thread.run(Thread.java:750)

```
</details>

--- a/compiler/src/dotty/tools/dotc/ast/Desugar.scala
+++ b/compiler/src/dotty/tools/dotc/ast/Desugar.scala
@@ -381,7 +381,7 @@ object desugar {
     tree match
       case untpd.Block(stats, expr) =>
         val (untpdTypeVariables, otherStats) = stats.span {
-          case tdef @ untpd.TypeDef(name, _) => name.isVarPattern
+          case tdef @ untpd.TypeDef(name, _) => !tdef.isBackquoted && name.isVarPattern
           case _ => false
         }
         val pattern = if otherStats.isEmpty then expr else untpd.cpy.Block(tree)(otherStats, expr)

--- a/compiler/src/dotty/tools/dotc/ast/Desugar.scala
+++ b/compiler/src/dotty/tools/dotc/ast/Desugar.scala
@@ -384,8 +384,11 @@ object desugar {
           case tdef @ untpd.TypeDef(name, _) => !tdef.isBackquoted && name.isVarPattern
           case _ => false
         }
+        val untpdCaseTypeVariables = untpdTypeVariables.asInstanceOf[List[untpd.TypeDef]].map {
+          tdef => tdef.withMods(tdef.mods | Case)
+        }
         val pattern = if otherStats.isEmpty then expr else untpd.cpy.Block(tree)(otherStats, expr)
-        (untpdTypeVariables.asInstanceOf[List[untpd.TypeDef]], pattern)
+        (untpdCaseTypeVariables, pattern)
       case _ =>
         (Nil, tree)
 

--- a/compiler/src/dotty/tools/dotc/quoted/QuotePatterns.scala
+++ b/compiler/src/dotty/tools/dotc/quoted/QuotePatterns.scala
@@ -15,6 +15,7 @@ import dotty.tools.dotc.core.NameKinds.PatMatGivenVarName
 import dotty.tools.dotc.core.Names.*
 import dotty.tools.dotc.core.StdNames.*
 import dotty.tools.dotc.core.Symbols.*
+import dotty.tools.dotc.core.TypeOps.*
 import dotty.tools.dotc.core.Types.*
 import dotty.tools.dotc.reporting.IllegalVariableInPatternAlternative
 import dotty.tools.dotc.transform.SymUtils._
@@ -23,6 +24,33 @@ import scala.collection.mutable
 
 object QuotePatterns:
   import tpd._
+
+  /** Check for restricted patterns */
+  def checkPattern(quotePattern: QuotePattern)(using Context): Unit = new tpd.TreeTraverser {
+    def traverse(tree: Tree)(using Context): Unit = tree match {
+      case _: SplicePattern =>
+      case tdef: TypeDef if tdef.symbol.isClass =>
+        val kind = if tdef.symbol.is(Module) then "objects" else "classes"
+        report.error(em"Implementation restriction: cannot match $kind", tree.srcPos)
+      case tree: NamedDefTree =>
+        if tree.name.is(NameKinds.WildcardParamName) then
+          report.warning(
+            "Use of `_` for lambda in quoted pattern. Use explicit lambda instead or use `$_` to match any term.",
+            tree.srcPos)
+        if tree.name.isTermName && !tree.nameSpan.isSynthetic && tree.name != nme.ANON_FUN && tree.name.startsWith("$") then
+          report.error("Names cannot start with $ quote pattern", tree.namePos)
+        traverseChildren(tree)
+      case _: Match =>
+        report.error("Implementation restriction: cannot match `match` expressions", tree.srcPos)
+      case _: Try =>
+        report.error("Implementation restriction: cannot match `try` expressions", tree.srcPos)
+      case _: Return =>
+        report.error("Implementation restriction: cannot match `return` statements", tree.srcPos)
+      case _ =>
+        traverseChildren(tree)
+    }
+
+  }.traverse(quotePattern.body)
 
   /** Encode the quote pattern into an `unapply` that the pattern matcher can handle.
    *

--- a/compiler/src/dotty/tools/dotc/typer/QuotesAndSplices.scala
+++ b/compiler/src/dotty/tools/dotc/typer/QuotesAndSplices.scala
@@ -326,6 +326,7 @@ object QuotesAndSplices {
       }
 
       override def transform(tree: Tree)(using Context) = tree match
+        // TODO: handle TypeBoundsTree, LambdaTypeTree as well as method parameters in DefTrees?
         case tree @ AppliedTypeTree(tpt, args) =>
           val args1: List[Tree] = args.zipWithConserve(tpt.tpe.typeParams.map(_.paramVarianceSign)) { (arg, v) =>
             arg.tpe match {

--- a/compiler/src/dotty/tools/dotc/typer/QuotesAndSplices.scala
+++ b/compiler/src/dotty/tools/dotc/typer/QuotesAndSplices.scala
@@ -156,8 +156,9 @@ trait QuotesAndSplices {
       typedSplicePattern(untpd.cpy.SplicePattern(tree)(splice.body, args), pt)
   }
 
-  /** Type a pattern variable name `t` in quote pattern as `${given t$giveni: Type[t @ _]}`.
-   *  The resulting pattern is the split in `splitQuotePattern`.
+  /** Type check a type binding reference in a quoted pattern.
+   *
+   *  If no binding exists with that name, this becomes the definition of a new type binding.
    */
   def typedQuotedTypeVar(tree: untpd.Ident, pt: Type)(using Context): Tree =
     val typeSymInfo = pt match
@@ -178,15 +179,13 @@ trait QuotesAndSplices {
         warnOnInferredBounds(typeSym)
         ref(typeSym)
       case None =>
-        val spliceContext = quotePatternSpliceContext
-        val name = tree.name.toTypeName
-        val nameOfSyntheticGiven = PatMatGivenVarName.fresh(tree.name.toTermName)
-        val expr = untpd.cpy.Ident(tree)(nameOfSyntheticGiven)
-        val typeSym = newSymbol(spliceContext.owner, name, EmptyFlags, typeSymInfo, NoSymbol, tree.span)
-        typeSym.addAnnotation(Annotation(New(ref(defn.QuotedRuntimePatterns_patternTypeAnnot.typeRef)).withSpan(tree.span)))
+        if ctx.mode.is(Mode.InPatternAlternative) then
+          report.error(IllegalVariableInPatternAlternative(tree.name), tree.srcPos)
+        val typeSym = inContext(quotePatternOuterContext(ctx)) {
+          newSymbol(ctx.owner, tree.name.toTypeName, Case, typeSymInfo, NoSymbol, tree.span)
+        }
         addQuotedPatternTypeVariable(typeSym)
-        val pat = typedPattern(expr, defn.QuotedTypeClass.typeRef.appliedTo(typeSym.typeRef))(using spliceContext)
-        pat.select(tpnme.Underlying)
+        Bind(typeSym, untpd.Ident(nme.WILDCARD).withType(typeSymInfo)).withSpan(tree.span)
 
   private def checkSpliceOutsideQuote(tree: untpd.Tree)(using Context): Unit =
     if (level == 0 && !ctx.owner.ownersIterator.exists(_.isInlineMethod))
@@ -199,234 +198,11 @@ trait QuotesAndSplices {
             |""", tree.srcPos
       )
 
-  /** Split a typed quoted pattern is split into its type bindings, pattern expression and inner patterns.
-   *  Type definitions with `@patternType` will be inserted in the pattern expression for each type binding.
-   *
-   *  A quote pattern
-   *  ```
-   *  case '{ type ${given t$giveni: Type[t @ _]}; ${ls: Expr[List[t]]} } => ...
-   *  ```
-   *  will return
-   *  ```
-   *  (
-   *    Map(<t$giveni>: Symbol -> <t>: Symbol),
-   *    <'{
-   *       @scala.internal.Quoted.patternType type t
-   *       scala.internal.Quoted.patternHole[List[t]]
-   *    }>: Tree,
-   *    List(<ls: Expr[List[t]]>: Tree)
-   *  )
-   *  ```
-   */
-  private def splitQuotePattern(quoted: Tree)(using Context): (SeqMap[Symbol, Symbol], Tree, List[Tree]) = {
-    val ctx0 = ctx
-
-    val bindSymMapping: SeqMap[Symbol, Symbol] = unapplyBindingsMapping(quoted)
-
-    object splitter extends tpd.TreeMap {
-      private var variance: Int = 1
-
-      inline private def atVariance[T](v: Int)(op: => T): T = {
-        val saved = variance
-        variance = v
-        val res = op
-        variance = saved
-        res
-      }
-
-      val patBuf = new mutable.ListBuffer[Tree]
-      val freshTypePatBuf = new mutable.ListBuffer[Tree]
-      val freshTypeBindingsBuff = new mutable.ListBuffer[Tree]
-      val typePatBuf = new mutable.ListBuffer[Tree]
-      override def transform(tree: Tree)(using Context) = tree match {
-        case Typed(splice @ SplicePattern(pat, Nil), tpt) if !tpt.tpe.derivesFrom(defn.RepeatedParamClass) =>
-          transform(tpt) // Collect type bindings
-          transform(splice)
-        case SplicePattern(pat, args) =>
-          val patType = pat.tpe.widen
-          val patType1 = patType.translateFromRepeated(toArray = false)
-          val pat1 = if (patType eq patType1) pat else pat.withType(patType1)
-          patBuf += pat1
-          if args.isEmpty then ref(defn.QuotedRuntimePatterns_patternHole.termRef).appliedToType(tree.tpe).withSpan(tree.span)
-          else ref(defn.QuotedRuntimePatterns_higherOrderHole.termRef).appliedToType(tree.tpe).appliedTo(SeqLiteral(args, TypeTree(defn.AnyType))).withSpan(tree.span)
-        case Select(pat: Bind, _) if tree.symbol.isTypeSplice =>
-          val sym = tree.tpe.dealias.typeSymbol
-          if sym.exists then
-            val tdef = TypeDef(sym.asType).withSpan(sym.span)
-            val nameOfSyntheticGiven = pat.symbol.name.toTermName
-            freshTypeBindingsBuff += transformTypeBindingTypeDef(nameOfSyntheticGiven, tdef, freshTypePatBuf)
-            TypeTree(tree.tpe.dealias).withSpan(tree.span)
-          else
-            tree
-        case tdef: TypeDef  =>
-          if tdef.symbol.hasAnnotation(defn.QuotedRuntimePatterns_patternTypeAnnot) then
-            transformTypeBindingTypeDef(PatMatGivenVarName.fresh(tdef.name.toTermName), tdef, typePatBuf)
-          else if tdef.symbol.isClass then
-            val kind = if tdef.symbol.is(Module) then "objects" else "classes"
-            report.error(em"Implementation restriction: cannot match $kind", tree.srcPos)
-            EmptyTree
-          else
-            super.transform(tree)
-        case tree @ AppliedTypeTree(tpt, args) =>
-            val args1: List[Tree] = args.zipWithConserve(tpt.tpe.typeParams.map(_.paramVarianceSign)) { (arg, v) =>
-              arg.tpe match {
-                case _: TypeBounds => transform(arg)
-                case _ => atVariance(variance * v)(transform(arg))
-              }
-            }
-            cpy.AppliedTypeTree(tree)(transform(tpt), args1)
-        case tree: NamedDefTree =>
-          if tree.name.is(NameKinds.WildcardParamName) then
-            report.warning(
-              "Use of `_` for lambda in quoted pattern. Use explicit lambda instead or use `$_` to match any term.",
-              tree.srcPos)
-          if tree.name.isTermName && !tree.nameSpan.isSynthetic && tree.name.startsWith("$") then
-            report.error("Names cannot start with $ quote pattern ", tree.namePos)
-          super.transform(tree)
-        case _: Match =>
-          report.error("Implementation restriction: cannot match `match` expressions", tree.srcPos)
-          EmptyTree
-        case _: Try =>
-          report.error("Implementation restriction: cannot match `try` expressions", tree.srcPos)
-          EmptyTree
-        case _: Return =>
-          report.error("Implementation restriction: cannot match `return` statements", tree.srcPos)
-          EmptyTree
-        case _ =>
-          super.transform(tree)
-      }
-
-      private def transformTypeBindingTypeDef(nameOfSyntheticGiven: TermName, tdef: TypeDef, buff: mutable.Builder[Tree, List[Tree]])(using Context): Tree = {
-        if ctx.mode.is(Mode.InPatternAlternative) then
-          report.error(IllegalVariableInPatternAlternative(tdef.symbol.name), tdef.srcPos)
-        if variance == -1 then
-          tdef.symbol.addAnnotation(Annotation(New(ref(defn.QuotedRuntimePatterns_fromAboveAnnot.typeRef)).withSpan(tdef.span)))
-        val bindingType = bindSymMapping(tdef.symbol).typeRef
-        val bindingTypeTpe = AppliedType(defn.QuotedTypeClass.typeRef, bindingType :: Nil)
-        val sym = newPatternBoundSymbol(nameOfSyntheticGiven, bindingTypeTpe, tdef.span, flags = ImplicitVal)(using ctx0)
-        buff += Bind(sym, untpd.Ident(nme.WILDCARD).withType(bindingTypeTpe)).withSpan(tdef.span)
-        super.transform(tdef)
-      }
-    }
-
-    val shape0 = splitter.transform(quoted)
-    val patterns = (splitter.typePatBuf.iterator ++ splitter.freshTypePatBuf.iterator ++ splitter.patBuf.iterator).toList
-    val freshTypeBindings = splitter.freshTypeBindingsBuff.result()
-
-    val shape1 = shape0 match
-      case Block(stats @ ((tdef: TypeDef) :: rest), expr) if tdef.symbol.hasAnnotation(defn.QuotedRuntimePatterns_patternTypeAnnot) =>
-        val (bindings, otherStats) = stats.span {
-          case tdef: TypeDef => tdef.symbol.hasAnnotation(defn.QuotedRuntimePatterns_patternTypeAnnot)
-          case _ => true
-        }
-        cpy.Block(shape0)(bindings ::: freshTypeBindings ::: otherStats, expr)
-      case _ =>
-        seq(freshTypeBindings, shape0)
-
-    val shape2 =
-      if (freshTypeBindings.isEmpty) shape1
-      else {
-        val isFreshTypeBindings = freshTypeBindings.map(_.symbol).toSet
-        val typeMap = new TypeMap() {
-          def apply(tp: Type): Type = tp match {
-            case tp: TypeRef if tp.symbol.isTypeSplice =>
-              val tp1 = tp.dealias
-              if (isFreshTypeBindings(tp1.typeSymbol)) tp1
-              else tp
-            case tp => mapOver(tp)
-          }
-        }
-        new TreeTypeMap(typeMap = typeMap).transform(shape1)
-      }
-
-    (bindSymMapping, shape2, patterns)
-  }
-
-  /** For each type variable defined in the quote pattern we generate an equivalent
-   *  binding that will be as type variable in the encoded `unapply` of the quote pattern.
-   *
-   *  @return Mapping from type variable symbols defined in the quote pattern into
-   *          type variable definitions for the `unapply` of the quote pattern.
-   *          This mapping retains the original type variable definition order.
-   */
-  private def unapplyBindingsMapping(quoted: Tree)(using Context): SeqMap[Symbol, Symbol] = {
-    val mapping = mutable.LinkedHashMap.empty[Symbol, Symbol]
-
-    // Collect all existing type variable bindings and create new symbols for them.
-    // The old info is used, it may contain references to the old symbols.
-    new tpd.TreeTraverser {
-      def traverse(tree: Tree)(using Context): Unit = tree match {
-        case _: SplicePattern =>
-        case Select(pat: Bind, _) if tree.symbol.isTypeSplice =>
-          val sym = tree.tpe.dealias.typeSymbol
-          if sym.exists then registerNewBindSym(sym)
-        case tdef: TypeDef  =>
-          if tdef.symbol.hasAnnotation(defn.QuotedRuntimePatterns_patternTypeAnnot) then
-            registerNewBindSym(tdef.symbol)
-          traverseChildren(tdef)
-        case _ =>
-          traverseChildren(tree)
-      }
-      private def registerNewBindSym(sym: Symbol): Unit =
-        if !mapping.contains(sym) then
-          mapping(sym) = newSymbol(ctx.owner, sym.name.toString.stripPrefix("$").toTypeName, Case | sym.flags, sym.info, coord = quoted.span)
-    }.traverse(quoted)
-
-    // Replace symbols in `mapping` in the infos of the new symbol and register GADT bounds.
-    // GADT bounds need to be added after the info is updated to avoid references to the old symbols.
-    var oldBindings: List[Symbol] = mapping.keys.toList
-    var newBindingsRefs: List[Type] = mapping.values.toList.map(_.typeRef)
-    for newBindings <- mapping.values do
-      newBindings.info = newBindings.info.subst(oldBindings, newBindingsRefs)
-      ctx.gadtState.addToConstraint(newBindings) // This must be preformed after the info has been updated
-
-    mapping
-  }
-
-  /** Type a quote pattern `case '{ <quoted> } =>` qiven the a current prototype. Typing the pattern
-   *  will also transform it into a call to `scala.internal.quoted.Expr.unapply`.
+  /** Type a quote pattern `case '{ <pattern> } =>` given the a current prototype. Typing the pattern
+   *  will create a QuotePattern tree.
    *
    *  Code directly inside the quote is typed as an expression using Mode.QuotedPattern. Splices
    *  within the quotes become patterns again and typed accordingly.
-   *
-   *  ```
-   *  case '{ ($ls: List[t]) } =>
-   *    // `t$giveni` is of type `Type[t]` for some unknown `t`
-   *    // `t$giveni` is implicitly available
-   *    // `ls` is of type `Expr[List[t]]`
-   *    '{ val h: $t = $ls.head  }
-   *  ```
-   *
-   *  For each type splice we will create a new type binding in the pattern match (`t @ _` in this case)
-   *  and a corresponding type in the quoted pattern as a hole (`@patternType type t` in this case).
-   *  All these generated types are inserted at the start of the quoted code.
-   *
-   *  After typing the tree will resemble
-   *
-   *  ```
-   *  case '{ type ${given t$giveni: Type[t @ _]}; ${ls: Expr[List[t]]} } => ...
-   *  ```
-   *
-   *  Then the pattern is _split_ into the expression contained in the pattern replacing the splices by holes,
-   *  and the patterns in the splices. All these are recombined into a call to `Matcher.unapply`.
-   *
-   *  ```
-   *  case scala.internal.quoted.Expr.unapply[
-   *          KList[t @ _, KNil], // Type binging definition
-   *          Tuple2[Type[t], Expr[List[t]]] // Typing the result of the pattern match
-   *        ](
-   *          Tuple2.unapply
-   *            [Type[t], Expr[List[t]]] //Propagated from the tuple above
-   *            (given t$giveni @ _, ls @ _: Expr[List[t]]) // from the spliced patterns
-   *        )(
-   *         '{ // Runtime quote Matcher.unapply uses to mach against. Expression directly inside the quoted pattern without the splices
-   *            @scala.internal.Quoted.patternType type t
-   *            scala.internal.Quoted.patternHole[List[t]]
-   *          },
-   *          true, // If there is at least one type splice. Used to instantiate the context with or without GADT constraints
-   *          x$2 // tasty.Reflection instance
-   *        ) => ...
-   *  ```
    */
   private def typedQuotePattern(tree: untpd.Quote, pt: Type, quotes: Tree)(using Context): Tree = {
     val quoted = tree.body
@@ -453,83 +229,56 @@ trait QuotesAndSplices {
         untpdTypeVariables.head.srcPos,
         "\n\nSIP-53: https://docs.scala-lang.org/sips/quote-pattern-type-variable-syntax.html")
 
-    val (typeTypeVariables, patternCtx) =
+    if ctx.mode.is(Mode.InPatternAlternative) then
+      for tpVar <- untpdTypeVariables do
+        report.error(IllegalVariableInPatternAlternative(tpVar.name), tpVar.srcPos)
+
+    val (typeTypeVariables, patternBlockCtx) =
       val quoteCtx = quotePatternContext(quoted.isType)
       if untpdTypeVariables.isEmpty then (Nil, quoteCtx)
       else typedBlockStats(untpdTypeVariables)(using quoteCtx)
+    val patternCtx = patternBlockCtx.addMode(if quoted.isType then Mode.QuotedTypePattern else Mode.QuotedExprPattern)
 
-    val quoted1 = inContext(patternCtx) {
+    val allTypeBindings = List.newBuilder[Bind]
+    for tpVar <- typeTypeVariables do
+      val sym = tpVar.symbol
+      sym.setFlag(Case)
+      allTypeBindings += Bind(sym, untpd.Ident(nme.WILDCARD).withType(sym.info)).withSpan(tpVar.span)
+
+    val body1 = inContext(patternCtx) {
       for typeVariable <- typeTypeVariables do
         addQuotedPatternTypeVariable(typeVariable.symbol)
 
-      val pattern =
-        if quoted.isType then typedType(quoted0, WildcardType)(using patternCtx)
-        else typedExpr(quoted0, WildcardType)
-
-      if untpdTypeVariables.isEmpty then pattern
-      else tpd.Block(typeTypeVariables, pattern)
+      if quoted.isType then typedType(quoted0, WildcardType)
+      else typedExpr(quoted0, WildcardType)
     }
 
-    val (bindSymMapping, shape, splices) = splitQuotePattern(quoted1)
+    val extractTypeBindings = new TreeMapWithVariance {
+      override def transform(tree: Tree)(using Context) = tree match
+        case pat: Bind if pat.isType =>
+          if inContravariantPosition then
+            pat.symbol.addAnnotation(Annotation(New(ref(defn.QuotedRuntimePatterns_fromAboveAnnot.typeRef)).withSpan(pat.span)))
+          allTypeBindings += pat
+          TypeTree(pat.symbol.typeRef).withSpan(pat.span)
 
-    class ReplaceBindings extends TypeMap() {
-      override def apply(tp: Type): Type = tp match {
-        case tp: TypeRef =>
-          val tp1 = if (tp.symbol.isTypeSplice) tp.dealias else tp
-          mapOver(bindSymMapping.get(tp1.typeSymbol).fold(tp)(_.typeRef))
-        case tp => mapOver(tp)
-      }
+        case Typed(splice: SplicePattern, tpt) if !tpt.tpe.derivesFrom(defn.RepeatedParamClass) =>
+          // We drop the type ascription because it is redundant, the SplicePattern contains the same type
+          transform(tpt) // Collect type bindings
+          splice
+        case _: SplicePattern =>
+          tree
+
+        case _ =>
+          super.transform(tree)
     }
-    val replaceBindings = new ReplaceBindings
-    val patType = defn.tupleType(splices.tpes.map(tpe => replaceBindings(tpe.widen)))
+    val body2 = extractTypeBindings.transform(body1)
 
-    val typeBinds = bindSymMapping.values.map(sym =>
-      Bind(sym, untpd.Ident(nme.WILDCARD).withType(sym.info)).withSpan(quoted.span)
-    ).toList
-    val typeBindingsTuple = tpd.hkNestedPairsTypeTree(typeBinds)
+    val quoteClass = if quoted.isTerm then defn.QuotedExprClass else defn.QuotedTypeClass
+    val pt1 = quoteClass.typeRef.appliedTo(body2.tpe)
 
-    val replaceBindingsInTree = new TreeMap {
-      private var bindMap = Map.empty[Symbol, Symbol]
-      override def transform(tree: tpd.Tree)(using Context): tpd.Tree =
-        tree match {
-          case tree: Bind =>
-            val sym = tree.symbol
-            val newInfo = replaceBindings(sym.info)
-            val newSym = newSymbol(sym.owner, sym.name, sym.flags, newInfo, sym.privateWithin, sym.coord)
-            bindMap += sym -> newSym
-            Bind(newSym, transform(tree.body)).withSpan(sym.span)
-          case _ =>
-            super.transform(tree).withType(replaceBindingsInType(tree.tpe))
-        }
-      private val replaceBindingsInType = new ReplaceBindings {
-        override def apply(tp: Type): Type = tp match {
-          case tp: TermRef => bindMap.get(tp.termSymbol).fold[Type](tp)(_.typeRef)
-          case tp => super.apply(tp)
-        }
-      }
-    }
-
-    val splicePat =
-      if splices.isEmpty then ref(defn.EmptyTupleModule.termRef)
-      else typed(untpd.Tuple(splices.map(x => untpd.TypedSplice(replaceBindingsInTree.transform(x)))).withSpan(quoted.span), patType)
-
-    val quoteClass = if (quoted.isTerm) defn.QuotedExprClass else defn.QuotedTypeClass
-    val quotedPattern =
-      if (quoted.isTerm) tpd.Quote(shape, Nil).select(nme.apply).appliedTo(quotes)
-      else ref(defn.QuotedTypeModule_of.termRef).appliedToTypeTree(shape).appliedTo(quotes)
-
-    val matchModule = if quoted.isTerm then defn.QuoteMatching_ExprMatch else defn.QuoteMatching_TypeMatch
-    val unapplySym = if quoted.isTerm then defn.QuoteMatching_ExprMatch_unapply else defn.QuoteMatching_TypeMatch_unapply
-    val unapplyFun = quotes.asInstance(defn.QuoteMatchingClass.typeRef).select(matchModule).select(unapplySym)
-
-    val encodedPattern = UnApply(
-      fun = unapplyFun.appliedToTypeTrees(typeBindingsTuple :: TypeTree(patType) :: Nil),
-      implicits = quotedPattern :: Nil,
-      patterns = splicePat :: Nil,
-      proto = quoteClass.typeRef.appliedTo(replaceBindings(quoted1.tpe)))
-
-    if ctx.reporter.hasErrors then encodedPattern
-    else QuotePatterns.decode(encodedPattern) // TODO type directly into the encoded version
+    val quotePattern = QuotePattern(allTypeBindings.result(), body2, quotes, pt1)
+    QuotePatterns.checkPattern(quotePattern)
+    quotePattern
   }
 }
 
@@ -550,8 +299,7 @@ object QuotesAndSplices {
   /** Context used to type the contents of a quote pattern */
   def quotePatternContext(isTypePattern: Boolean)(using Context): Context =
     quoteContext.fresh.setNewScope
-      .addMode(if isTypePattern then Mode.QuotedTypePattern else Mode.QuotedExprPattern)
-      .retractMode(Mode.Pattern) // TODO do we need Mode.QuotedPattern?
+      .retractMode(Mode.Pattern)
       .setProperty(TypeVariableKey, collection.mutable.Map.empty)
 
   /** Context used to type the contents of a quote pattern splice */
@@ -559,10 +307,35 @@ object QuotesAndSplices {
     spliceContext
       .retractMode(Mode.QuotedPatternBits)
       .addMode(Mode.Pattern)
-      .withOwner(quotePatternOwner(ctx))
+      .withOwner(quotePatternOuterContext(ctx).owner)
 
-  /** First outer context owner that is outside of a quoted pattern context. */
-  private def quotePatternOwner(ctx: Context): Symbol =
-    if ctx.mode.isQuotedPattern then quotePatternOwner(ctx.outer) else ctx.owner
+  /** First outer context that is outside of a quoted pattern. */
+  def quotePatternOuterContext(ctx: Context): Context =
+    if ctx.mode.isQuotedPattern then quotePatternOuterContext(ctx.outer) else ctx
 
+  private[QuotesAndSplices] class TreeMapWithVariance extends TreeMap:
+      private var variance: Int = 1
+
+      def inContravariantPosition: Boolean = variance == -1
+
+      inline private def atVariance[T](v: Int)(op: => T): T = {
+        val saved = variance
+        variance = v
+        val res = op
+        variance = saved
+        res
+      }
+
+      override def transform(tree: Tree)(using Context) = tree match
+        case tree @ AppliedTypeTree(tpt, args) =>
+          val args1: List[Tree] = args.zipWithConserve(tpt.tpe.typeParams.map(_.paramVarianceSign)) { (arg, v) =>
+            arg.tpe match {
+              case _: TypeBounds => transform(arg)
+              case _ => atVariance(v * variance)(transform(arg))
+            }
+          }
+          cpy.AppliedTypeTree(tree)(transform(tpt), args1)
+        case _ =>
+          super.transform(tree)
+    end TreeMapWithVariance
 }

--- a/compiler/src/dotty/tools/dotc/typer/QuotesAndSplices.scala
+++ b/compiler/src/dotty/tools/dotc/typer/QuotesAndSplices.scala
@@ -242,7 +242,6 @@ trait QuotesAndSplices {
     val allTypeBindings = List.newBuilder[Bind]
     for tpVar <- typeTypeVariables do
       val sym = tpVar.symbol
-      sym.setFlag(Case)
       allTypeBindings += Bind(sym, untpd.Ident(nme.WILDCARD).withType(sym.info)).withSpan(tpVar.span)
 
     val body1 = inContext(patternCtx) {

--- a/tests/neg-macros/i16522.check
+++ b/tests/neg-macros/i16522.check
@@ -1,0 +1,7 @@
+-- [E007] Type Mismatch Error: tests/neg-macros/i16522.scala:10:45 -----------------------------------------------------
+10 |      case '{HCons($h1: hd1, HCons($h2: hd2, $_ : tl))} => '{$h1.toString ++ $h2.toString} // error
+   |                                             ^^^^^^^
+   |                                             Found:    tl
+   |                                             Required: HList
+   |
+   | longer explanation available when compiling with `-explain`

--- a/tests/neg-macros/i16522.scala
+++ b/tests/neg-macros/i16522.scala
@@ -1,0 +1,16 @@
+  import scala.quoted.*
+
+  sealed trait HList
+  case class HCons[+HD, TL <: HList](hd: HD, tl: TL) extends HList
+  case object HNil extends HList
+
+  def showFirstTwoImpl(e: Expr[HList])(using Quotes): Expr[String] = {
+    e match {
+      case '{HCons($h1, HCons($h2, $_))} => '{$h1.toString ++ $h2.toString}
+      case '{HCons($h1: hd1, HCons($h2: hd2, $_ : tl))} => '{$h1.toString ++ $h2.toString} // error
+      case '{HCons[hd, HCons[sd, tl]]($h1, HCons($h2, $_))} => '{$h1.toString ++ $h2.toString}
+      case _ => '{""}
+    }
+  }
+
+  transparent inline def showFirstTwo(inline xs: HList) = ${ showFirstTwoImpl('xs) }

--- a/tests/neg-macros/i6997b.scala
+++ b/tests/neg-macros/i6997b.scala
@@ -5,7 +5,7 @@ import scala.quoted.*
 inline def mcr(x: => Any): Any = ${mcrImpl('x)}
 
 def mcrImpl(body: Expr[Any])(using ctx: Quotes): Expr[Any] = {
-  val '{$x: $t} = body // error
+  val '{$x: $t} = body // error // error
   '{
     val tmp: $t = $x.asInstanceOf[$t] // error // error
     println(tmp)

--- a/tests/pos-macros/i14708.scala
+++ b/tests/pos-macros/i14708.scala
@@ -1,0 +1,10 @@
+import scala.quoted.*
+
+object Main {
+  def foo(a: Expr[Any])(using Quotes) = {
+    a match {
+      case '{ ($x: Set[t]).toSet } =>
+      case _ =>
+    }
+  }
+}

--- a/tests/pos-macros/i16522.scala
+++ b/tests/pos-macros/i16522.scala
@@ -1,0 +1,16 @@
+  import scala.quoted.*
+
+  sealed trait HList
+  case class HCons[+HD, TL <: HList](hd: HD, tl: TL) extends HList
+  case object HNil extends HList
+
+  def showFirstTwoImpl(e: Expr[HList])(using Quotes): Expr[String] = {
+    e match {
+      case '{HCons($h1, HCons($h2, $_))} => '{$h1.toString ++ $h2.toString}
+      case '{type tl <: HList; HCons($h1: hd1, HCons($h2: hd2, $_ : tl))} => '{$h1.toString ++ $h2.toString} // error
+      case '{HCons[hd, HCons[sd, tl]]($h1, HCons($h2, $_))} => '{$h1.toString ++ $h2.toString}
+      case _ => '{""}
+    }
+  }
+
+  transparent inline def showFirstTwo(inline xs: HList) = ${ showFirstTwoImpl('xs) }

--- a/tests/pos-macros/i18125.scala
+++ b/tests/pos-macros/i18125.scala
@@ -1,0 +1,10 @@
+import scala.quoted.*
+
+final class Foo[T](ns: T)
+
+def foo(using Quotes)(x: Expr[Any]): Unit =
+  x match
+    case '{ new Foo($y: b) } =>
+    case '{ new Foo($y: List[b]) } =>
+    case '{ type b; new Foo($y: b) } =>
+

--- a/tests/pos-macros/i18125b.scala
+++ b/tests/pos-macros/i18125b.scala
@@ -1,0 +1,23 @@
+package oolong.phobos
+
+import scala.quoted.*
+import scala.compiletime.*
+import scala.annotation.StaticAnnotation
+
+final class xmlns[T](ns: T) extends StaticAnnotation
+trait Namespace[T]{
+    val getNamespace: String
+}
+
+object common{
+  private def extractFeildNamespace(using Quotes)(
+      fieldAnnotations: List[Expr[Any]],
+  ): Expr[Option[String]] = {
+    import quotes.reflect.*
+
+    fieldAnnotations.collect { case '{ xmlns($namespace: b) } =>
+      '{ Some(summonInline[Namespace[b]].getNamespace) }
+    }
+    ???
+  }
+}

--- a/tests/pos-macros/i18250.scala
+++ b/tests/pos-macros/i18250.scala
@@ -1,0 +1,6 @@
+import scala.quoted.*
+
+def test(x: Expr[Any])(using Quotes): Unit =
+  x match
+    case '{ type t; type u <: t; () } =>
+    case '{ type t <: Comparable[t]; () } =>

--- a/tests/pos-macros/mirrorQuotePattern.scala
+++ b/tests/pos-macros/mirrorQuotePattern.scala
@@ -1,0 +1,13 @@
+import scala.deriving._
+import scala.quoted._
+
+private def derivedExpr[T](mirrorExpr: Expr[Mirror.Of[T]])(using Quotes, Type[T]): Expr[Any] = {
+  mirrorExpr match {
+    case '{ $mirrorExpr : Mirror.Sum { type MirroredElemTypes = mirroredElemTypes } } =>
+      '{ liftableSum[mirroredElemTypes]($mirrorExpr) }
+    case '{ type mirroredElemTypes; $mirrorExpr : Mirror.Sum { type MirroredElemTypes = mirroredElemTypes } } =>
+      '{ liftableSum[mirroredElemTypes]($mirrorExpr) }
+  }
+}
+
+def liftableSum[MElemTypes](mirror: Mirror.Sum { type MirroredElemTypes = MElemTypes }): Any = ???

--- a/tests/pos-macros/mirrorQuotePattern2.scala
+++ b/tests/pos-macros/mirrorQuotePattern2.scala
@@ -1,0 +1,10 @@
+import scala.deriving._
+import scala.quoted._
+
+private def derivedExpr(x: Expr[Any])(using Quotes): Unit =
+  x match
+    case '{ type mtp1; ($m1 : Mirror.Sum { type MirroredElemTypes = mtp1 } & Mirror.Of[Any], $m2 : Mirror.Sum { type MirroredElemTypes = mtp2 } & Mirror.Of[Any]); ??? } =>
+      val _: Expr[Mirror.Sum { type MirroredElemTypes = mtp1 } & Mirror.Of[Any]] = m1
+      val _: Expr[Mirror.Sum { type MirroredElemTypes = mtp2 } & Mirror.Of[Any]] = m2
+      '{ $m1: Mirror.Sum { type MirroredElemTypes = mtp1 } }
+      '{ $m2: Mirror.Sum { type MirroredElemTypes = mtp2 } }

